### PR TITLE
[Enhancement] Optimize delete_unused_rowset, make system call out of lock

### PR DESCRIPTION
**Pull request by meegoo**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

1. Make system call out of lock
2. Smooth unlink process, avoid too much pressure to filesystem meta

### CODE REVIEW

File: be/src/storage/storage_engine.cpp [hunk 1]
```diff
@@ -915,19 +915,41 @@ Status StorageEngine::_do_sweep(const std::string& scan_root, const time_t& loca
...
+        for (auto it = _unused_rowsets.begin(); it != _unused_rowsets.end();) {
+            if (it->second.use_count() != 1) {
+                ++it;
+            } else if (it->second->need_delete_file()) {
+                delete_rowsets.emplace_back(it->second);
+                it = _unused_rowsets.erase(it);
+            }
+            if (delete_rowsets.size() > batch_size) {
+                auto new_interval = *interval / (_unused_rowsets.size() + batch_size / batch_size);
+                *interval = new_interval > 0 ? new_interval : 1;
```
Comment (imay) [inline be/src/storage/storage_engine.cpp:935]: And I think it is not a good way to change interval in this function.
you can return `number_removed_rowsests`, and let caller decide what will do next

File: be/src/storage/storage_engine.cpp [hunk 2]
```diff
@@ -915,19 +915,41 @@ Status StorageEngine::_do_sweep(const std::string& scan_root, const time_t& loca
...
+        std::lock_guard lock(_gc_mutex);
+        for (auto it = _unused_rowsets.begin(); it != _unused_rowsets.end();) {
+            if (it->second.use_count() != 1) {
+                ++it;
+            } else if (it->second->need_delete_file()) {
+                delete_rowsets.emplace_back(it->second);
+                it = _unused_rowsets.erase(it);
+            }
+            if (delete_rowsets.size() > batch_size) {
+                auto new_interval = *interval / (_unused_rowsets.size() + batch_size / batch_size);
```
Comment (chaoyli) [inline be/src/storage/storage_engine.cpp:934]: _unused_rowsets.size() + batch_size / batch_size is equals to _unused_rowsets.size() + 1

File: be/src/storage/storage_engine.cpp [hunk 3]
```diff
@@ -915,19 +915,41 @@ Status StorageEngine::_do_sweep(const std::string& scan_root, const time_t& loca
...
-            VLOG(3) << "start to remove rowset:" << it->second->rowset_id()
-                    << ", version:" << it->second->version().first << "-" << it->second->version().second;
-            Status status = it->second->remove();
-            VLOG(3) << "remove rowset:" << it->second->rowset_id() << " finished. status:" << status;
-            it = _unused_rowsets.erase(it);
+void StorageEngine::delete_unused_rowset(int32_t* deleted_pct) {
+    MonotonicStopWatch timer;
+    timer.start();
+    std::vector<RowsetSharedPtr> delete_rowsets;
+    int64_t batch_size = 4096;
```
Comment (imay) [inline be/src/storage/storage_engine.cpp:922]:
```suggestion
    constexpr int64_t batch_size = 4096;
```

File: be/src/storage/storage_engine.cpp [hunk 4]
```diff
@@ -915,19 +915,41 @@ Status StorageEngine::_do_sweep(const std::string& scan_root, const time_t& loca
...
-    for (auto it = _unused_rowsets.begin(); it != _unused_rowsets.end();) {
-        if (it->second.use_count() != 1) {
-            ++it;
-        } else if (it->second->need_delete_file()) {
-            VLOG(3) << "start to remove rowset:" << it->second->rowset_id()
-                    << ", version:" << it->second->version().first << "-" << it->second->version().second;
-            Status status = it->second->remove();
-            VLOG(3) << "remove rowset:" << it->second->rowset_id() << " finished. status:" << status;
-            it = _unused_rowsets.erase(it);
+void StorageEngine::delete_unused_rowset(int32_t* deleted_pct) {
```
Comment (imay) [inline be/src/storage/storage_engine.cpp:918]: why not return double?
```suggestion
double StorageEngine::delete_unused_rowset() {
```

File: be/src/storage/storage_engine.cpp [hunk 5]
```diff
@@ -915,19 +915,41 @@ Status StorageEngine::_do_sweep(const std::string& scan_root, const time_t& loca
...
+        std::lock_guard lock(_gc_mutex);
+        for (auto it = _unused_rowsets.begin(); it != _unused_rowsets.end();) {
+            if (it->second.use_count() != 1) {
+                ++it;
+            } else if (it->second->need_delete_file()) {
+                delete_rowsets.emplace_back(it->second);
+                it = _unused_rowsets.erase(it);
+            }
+            if (delete_rowsets.size() >= batch_size) {
+                *deleted_pct = 100 * batch_size / (_unused_rowsets.size() + batch_size);
```
Comment (imay) [inline be/src/storage/storage_engine.cpp:935]: if the _unused_rowsets is very very large, *deleted_pct will be 0.
And this will cause a div 0 crash.

**Review by trueeyu — APPROVED** [review]

**Review by chaoyli — APPROVED** [review]

**Pull request merged**